### PR TITLE
Cluster watching

### PR DESCRIPTION
**Pull request by philbrookes**

### CODE REVIEW

File: .gitignore
Comment (maleck13) [inline .gitignore:17]: Prob can just copy this over?
https://github.com/kcp-dev/kcp-glbc/blob/main/.gitignore

File: Dockerfile
Comment (maleck13) [inline Dockerfile:13]: same here? https://github.com/kcp-dev/kcp-glbc/blob/main/Dockerfile

File: Makefile
Comment (maleck13) [inline Makefile:68]: do we want to add the basics like make lint make test etc? not essential

File: config/rbac/secret_editor_role.yaml
Comment (maleck13) [inline config/rbac/secret_editor_role.yaml:15]: this seems wrong?

File: config/rbac/secret_viewer_role.yaml
Comment (maleck13) [inline config/rbac/secret_viewer_role.yaml:15]: same here some kind of weird generation perhaps?

File: config/samples/secret.yaml
Comment (maleck13) [inline config/samples/secret.yaml:5]: just a thought for the future but prob should add our own label to secrets to indicate which clusters users want us to manage ingress for?

File: pkg/_internal/metadata/labels_test.go
Comment (maleck13) [inline pkg/_internal/metadata/labels_test.go:1]: assuming most of this was copied over?

File: pkg/controllers/secret/secret_controller.go
Comment (maleck13) [inline pkg/controllers/secret/secret_controller.go:2]: we should prob remove this header and set it similar to the one in KCP
``` Copyright 2021 The MultiCluster Traffic Controller Authors.``` and leave the apache license part

File: pkg/controllers/secret/secret_controller.go
Comment (maleck13) [inline pkg/controllers/secret/secret_controller.go:65]: Yeah I think, while we can support the format we should prob have our own label, WDYT? it can be a follow on issue though

File: pkg/controllers/secret/secret_controller.go
Comment (maleck13) [inline pkg/controllers/secret/secret_controller.go:92]: is it possible to add a filter, so we only see secrets with these labels?

File: pkg/controllers/secret/secret_controller.go
Comment (maleck13) [inline pkg/controllers/secret/secret_controller.go:56]: I think we could change this also to be ClusterConfig. But again not needed right now

File: pkg/controllers/secret/secret_controller.go
Comment (maleck13) [inline pkg/controllers/secret/secret_controller.go:99]: any reason this couldn't be unmarshalled into a rest.Config. All the fields seem to match and then just need to set the host?

File: pkg/controllers/secret/secret_controller.go
Comment (maleck13) [inline pkg/controllers/secret/secret_controller.go:113]: see above about rest.Config

File: config/rbac/secret_editor_role.yaml
Comment (philbrookes) [reply config/rbac/secret_editor_role.yaml:15]: yeah, this is left over from using kubebuilder to make the controller boilerplate, I'll remove it.

File: config/samples/secret.yaml
Comment (philbrookes) [reply config/samples/secret.yaml:5]: I did it this way, as this is how Argo labels it's secrets for clusters. Would we expect a user of Argo to go and add our label to the secrets created by Argo?

File: pkg/_internal/metadata/labels_test.go
Comment (philbrookes) [reply pkg/_internal/metadata/labels_test.go:1]: yes

File: pkg/controllers/secret/secret_controller.go
Comment (philbrookes) [reply pkg/controllers/secret/secret_controller.go:2]: 👍

File: pkg/multiClusterWatch/main.go
Comment (maleck13) [inline pkg/multiClusterWatch/main.go:145]: I am guessing start is called in it own go routine

Comment (maleck13) [reply pkg/multiClusterWatch/main.go:145]: have we tried with multiple secrets yet?

**Review by maleck13 — APPROVED** [review]
mostly looks fine. Just a few comments that might be worth addressing before merging

File: config/samples/secret.yaml
Comment (maleck13) [reply config/samples/secret.yaml:5]: I think this is not a bit problem from my understanding you can create the secrets declaratively but not an issue right now. I think though we would also want them to add annotations such as geo and cloud provider at somepoint

File: PROJECT
Comment (mikenairn) [inline PROJECT:15]: I'm not sure any of this should be added since we don't provide that api?

File: cmd/multi-cluster-traffic-controller/main.go
Comment (mikenairn) [inline cmd/multi-cluster-traffic-controller/main.go:91]: Is there a reason we are moving main.go here? kubebuilder scaffold commands seem to not like it:
```
kubebuilder create api --group kuadrant.io --version v1 --kind DNSRecord
...
2023/01/06 10:32:36 failed to create API: unable to run pre-scaffold tasks of "base.go.kubebuilder.io/v3": main.go file should present in the root directory
```

File: config/samples/secret.yaml
Comment (david-martin) [reply config/samples/secret.yaml:5]: It sounds reasonable for the controller to only look for secrets with a specific label for mctc.
It would avoid a dependency on the labelling that ArgoCD expects and allow some clusters to be omitted from mctc but still used by ArgoCD

File: Dockerfile
Comment (philbrookes) [reply Dockerfile:13]: I'm not seeing the value of that

File: PROJECT
Comment (philbrookes) [reply PROJECT:15]: nice spot. Removed.

File: pkg/multiClusterWatch/main.go
Comment (philbrookes) [reply pkg/multiClusterWatch/main.go:145]: Not yet. This does need testing, but yes, the kubebuilder manager runs this in a separate goroute.

File: cmd/multi-cluster-traffic-controller/main.go
Comment (philbrookes) [reply cmd/multi-cluster-traffic-controller/main.go:91]: I've moved it back now, it was just for personal preference

File: pkg/controllers/secret/secret_controller.go
Comment (philbrookes) [reply pkg/controllers/secret/secret_controller.go:92]: I'm struggling with this in kubebuilder. So I'll merge this PR now, to unblock other work, and keep looking into this for a separate PR.

File: .gitignore
Comment (mikenairn) [inline .gitignore:18]: Not sure why we need this, but also not a big deal to add it here i guess

File: .gitignore
Comment (mikenairn) [inline .gitignore:21]: Sounds fancy "temporal directoryo"

**Review by mikenairn — APPROVED** [review]

**Pull request merged**
